### PR TITLE
108341 extend create comment to send mail

### PR DESCRIPTION
**Pull request by janid1967**
[AB#108341](https://statoildeveloper.visualstudio.com/b6d97460-2e53-47ae-aad4-b41396d90828/_workitems/edit/108341)

### CODE REVIEW

File: src/Equinor.ProCoSys.Completion.Command/Comments/CommentService.cs [hunk 1]
```diff
@@ -15,29 +16,37 @@ namespace Equinor.ProCoSys.Completion.Command.Comments;
...
+        _completionMailService = completionMailService;
+        _deepLinkUtility = deepLinkUtility;
         _logger = logger;
     }
 
     public async Task<CommentDto> AddAndSaveAsync(
         IUnitOfWork unitOfWork,
-        string parentType,
-        Guid parentGuid,
+        IEntityContext parentEntity,
```
Comment (jsoi-eq) [inline src/Equinor.ProCoSys.Completion.Command/Comments/CommentService.cs:37]: I'm a bit unsure about IEntityContext.
It feels we are adding more and more interfaces to implement on the domain objects.
And this seems to be related specifically to email, should we have email specific functionality in (i.e. GetEmailContext) in a domain object?

Perhaps it is possible to achieve this using extension method instead?
I hacked together something that seems to work and can show you on Monday.

Comment (jsoi-eq) [reply src/Equinor.ProCoSys.Completion.Command/Comments/CommentService.cs:37]: Suggested extension methods:
public static class EmailContextExtension
{
    public static string GetContextType(this IHaveGuid aggregateRoot) => aggregateRoot.GetType().Name;

    public static dynamic GetEmailContext(this IHaveGuid aggregateRoot)
    {
        dynamic emailContext = new ExpandoObject();
        emailContext.Entity = aggregateRoot;
        return emailContext;
    }
}

File: src/Equinor.ProCoSys.Completion.Command/Email/DeepLinkUtility.cs [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    public DeepLinkUtility(IPlantProvider plantProvider, IOptionsMonitor<ApplicationOptions> options)
+    {
+        _plantProvider = plantProvider;
+        _options = options;
+    }
+
+    public string CreateUrl(string contextName, Guid guid)
+    {
+        var plant = _plantProvider.Plant[4..];
+        switch (contextName)
```
[the remark this answers is not in the document]
Comment (janid1967) [reply src/Equinor.ProCoSys.Completion.Command/Email/DeepLinkUtility.cs:23]: not as I'm aware of

File: src/Equinor.ProCoSys.Completion.Command/Email/DeepLinkUtility.cs [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+    public DeepLinkUtility(IPlantProvider plantProvider, IOptionsMonitor<ApplicationOptions> options)
+    {
+        _plantProvider = plantProvider;
+        _options = options;
+    }
+
+    public string CreateUrl(string contextName, Guid guid)
+    {
+        var plant = _plantProvider.Plant[4..];
```
Comment (jsoi-eq) [inline src/Equinor.ProCoSys.Completion.Command/Email/DeepLinkUtility.cs:22]: Should we extract this line to a method with a name to better explain why we are doing a substring here?

File: src/tests/Equinor.ProCoSys.Completion.Domain.Tests/IHaveGuidExtensionTests.cs [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+        // Assert
+        Assert.AreEqual(_dut, emailContext.Entity);
+        Assert.AreEqual("Test", emailContext.MyProp);
+    }
+
+    private class TestableEntity : IHaveGuid
+    {
+        public Guid Guid { get; } = Guid.NewGuid();
+        // ReSharper disable once UnusedMember.Local
+        public string GetContextType() => nameof(TestableEntity);
```
Comment (jsoi-eq) [inline src/tests/Equinor.ProCoSys.Completion.Domain.Tests/IHaveGuidExtensionTests.cs:38]: Shouldn't these methods be in the extension class and not the TestableEntity class?
And you can add a property to the TestableEntity class in order to do the assert?

**Review by jsoi-eq — COMMENTED** [review]
Awaiting feedback

File: src/tests/Equinor.ProCoSys.Completion.Command.Tests/Comments/CommentServiceTests.cs [hunk 5]
```diff
@@ -159,4 +202,11 @@ public void AddAsync_ShouldAddCommentToRepository_WithMentions()
         Assert.IsTrue(_commentAddedToRepository.Mentions.Any(p => p.Guid == personB.Guid));
     }
     #endregion
+
+    private class TestableEntity : IHaveGuid
+    {
+        public Guid Guid { get; } = Guid.NewGuid();
+        public string GetContextType() => nameof(TestableEntity);
```
[the remark this answers is not in the document]
Comment (janid1967) [reply src/tests/Equinor.ProCoSys.Completion.Command.Tests/Comments/CommentServiceTests.cs]: rush job from my side ... good catch

File: src/tests/Equinor.ProCoSys.Completion.Domain.Tests/IHaveGuidExtensionTests.cs (hunk 4 above)
Comment (janid1967) [reply src/tests/Equinor.ProCoSys.Completion.Domain.Tests/IHaveGuidExtensionTests.cs]: rush job from my side ... good catch

**Review by jsoi-eq — APPROVED** [review]

**Pull request merged**